File: internal/analyzer/detectors/arp_spoof.py
from typing import Dict, List, Set

import pandas as pd

from internal.analyzer.base import Alert, BaseDetector, Level


class ARPSpoofDetector(BaseDetector):
    def __init__(self):
        super().__init__()

    def analyze(self, df: pd.DataFrame) -> List[Alert]:
        alerts = []
        if df.empty or "src_ip" not in df.columns or "src_mac" not in df.columns:
            return alerts

        ignored_macs = ["ff:ff:ff:ff:ff:ff", "00:00:00:00:00:00", ""]
        clean_df = df.dropna(subset=["src_ip", "src_mac"])
        clean_df = clean_df[
            (clean_df["src_ip"].str.strip() != "")
            & (~clean_df["src_mac"].str.lower().isin(ignored_macs))
        ]

        if clean_df.empty:
            return alerts

        grouped = clean_df.groupby("src_ip")["src_mac"].nunique()
        spoofed_ips = grouped[grouped > 1].index.tolist()

        for ip in spoofed_ips:
            macs = sorted(
                clean_df[clean_df["src_ip"] == ip]["src_mac"].unique().tolist()
            )
            latest_timestamp = clean_df[clean_df["src_ip"] == ip]["timestamp"].max()

            if self._should_alert("ARP_SPOOF", ip):
                alerts.append(
                    Alert(
                        timestamp=str(latest_timestamp),
                        level=Level.CRITICAL,
                        type="ARP_SPOOF",
                        message=f"Address conflict: IP {ip} seen on different devices: {', '.join(macs)}",
                        src_ip=ip,
                    )
                )

        return alerts

File: internal/analyzer/detectors/arp_spoof.go
package detectors

import (
	"hattic/internal/analyzer"
	"strings"
	"sync"
	"time"
)

type ARPSpoofDetector struct {
	*analyzer.BaseDetector
	ipToMacs map[string]map[string]bool
	mu       sync.Mutex
}

func NewARPSpoofDetector() *ARPSpoofDetector {
	return &ARPSpoofDetector{
		BaseDetector: analyzer.NewBaseDetector(5 * time.Minute),
		ipToMacs:     make(map[string]map[string]bool),
	}
}

func (d *ARPSpoofDetector) Analyze(packet *analyzer.PacketInfo) []analyzer.Alert {
	var alerts []analyzer.Alert

	if packet.SrcIP == "" || packet.SrcMAC == "" {
		return alerts
	}

	macLower := strings.ToLower(packet.SrcMAC)
	if macLower == "ff:ff:ff:ff:ff:ff" || macLower == "00:00:00:00:00:00" {
		return alerts
	}

	d.mu.Lock()
	defer d.mu.Unlock()

	if _, exists := d.ipToMacs[packet.SrcIP]; !exists {
		d.ipToMacs[packet.SrcIP] = make(map[string]bool)
	}
	d.ipToMacs[packet.SrcIP][macLower] = true

	if len(d.ipToMacs[packet.SrcIP]) > 1 {
		if d.ShouldAlert("ARP_SPOOF", packet.SrcIP) {
			macs := make([]string, 0, len(d.ipToMacs[packet.SrcIP]))
			for m := range d.ipToMacs[packet.SrcIP] {
				macs = append(macs, m)
			}
			alerts = append(alerts, analyzer.Alert{
				Timestamp: packet.Timestamp.Format(time.RFC3339),
				Level:     analyzer.LevelCritical,
				Type:      "ARP_SPOOF",
				Message:   "Address conflict: IP " + packet.SrcIP + " seen on different devices: " + strings.Join(macs, ", "),
				SrcIP:     packet.SrcIP,
			})
		}
	}

	return alerts
}
